### ISSUE TITLE
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

### ISSUE THREAD

**Issue by JonathanMatthey** [bundler/bundler#3372]
――― ERROR REPORT TEMPLATE ―――――――――――――――――――――――――――――――――――――――――――――――――――――――
- What did you do?
- What did you expect to happen?
- What happened instead?

Error details

    Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    Permission denied @ dir_s_mkdir - ./.gem.20150131-61502-djrp3

Gem files will remain installed in /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/nokogiri-1.6.6.2 for inspection.
Results logged to /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/extensions/x86_64-darwin-14/2.1.0-static/nokogiri-1.6.6.2/gem_make.out

    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/tmpdir.rb:85:in `mkdir'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/tmpdir.rb:85:in `block in mktmpdir'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/tmpdir.rb:142:in `create'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/tmpdir.rb:85:in `mktmpdir'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/ext_conf_builder.rb:14:in `build'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:161:in `block (2 levels) in build_extension'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:160:in `chdir'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:160:in `block in build_extension'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:159:in `synchronize'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:159:in `build_extension'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:198:in `block in build_extensions'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:195:in `each'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/ext/builder.rb:195:in `build_extensions'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:1436:in `block in build_extensions'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/user_interaction.rb:45:in `use_ui'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:1434:in `build_extensions'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/stub_specification.rb:60:in `build_extensions'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/basic_specification.rb:56:in `contains_requirable_file?'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:925:in `block in find_inactive_by_path'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:924:in `each'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:924:in `find'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/specification.rb:924:in `find_inactive_by_path'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems.rb:185:in `try_activate'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:132:in `rescue in require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:144:in `require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendor/net/http/persistent.rb:12:in `<top (required)>'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendored_persistent.rb:11:in `<top (required)>'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/fetcher.rb:1:in `<top (required)>'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/cli/install.rb:72:in `run'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/cli.rb:157:in `install'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendor/thor/command.rb:27:in `run'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendor/thor/invocation.rb:121:in `invoke_command'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendor/thor.rb:363:in `dispatch'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/vendor/thor/base.rb:440:in `start'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/cli.rb:10:in `start'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/bin/bundle:20:in `block in <top (required)>'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/lib/bundler/friendly_errors.rb:6:in `with_friendly_errors'
    /Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/bundler-1.8.0.pre/bin/bundle:18:in `<top (required)>'
    /Users/ratciti/.rbenv/versions/2.1.2/bin/bundle:23:in `load'
    /Users/ratciti/.rbenv/versions/2.1.2/bin/bundle:23:in `<main>'

Environment

    Bundler   1.8.0.pre
    Rubygems  2.2.2
    Ruby      2.1.2p95 (2014-05-08 revision 45877) [x86_64-darwin14.0]
    Git       1.9.3 (Apple Git-50)
――― TEMPLATE END ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/bundler/bundler/search?q=ERROR%3A+Failed+to+build+gem+native+extension.%0A%0A++++Permission+denied+%40+dir_s_mkdir+-+.%2F.gem.20150131-61502-djrp3%0A%0AGem+files+will+remain+installed+in+%2FUsers%2Fratciti%2F.rbenv%2Fversions%2F2.1.2%2Flib%2Fruby%2Fgems%2F2.1.0%2Fgems%2Fnokogiri-1.6.6.2+for+inspection.%0AResults+logged+to+%2FUsers%2Fratciti%2F.rbenv%2Fversions%2F2.1.2%2Flib%2Fruby%2Fgems%2F2.1.0%2Fextensions%2Fx86_64-darwin-14%2F2.1.0-static%2Fnokogiri-1.6.6.2%2Fgem_make.out%0A&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/bundler/bundler/issues/new

**Comment by supertopher** [comment]
TLDR:
* i do not believe this is an issue with bundler

this looks like a permissions error.  Bundler expects that you have permission to write to the directory the error:
 ``` Permission denied @ dir_s_mkdir - ./.gem.20150131-61502-djrp3 ``` 
as well as an error that came from ``` mkdir ``` both suggest incorrect permissions.  If you are on OSX user this is probably due to the inappropriate use of sudo.  OSX by convention doesn't use sudo for much of any CLI installation unless you are using Macports

if you are using brew try ``` brew doctor```  for help.  or uninstall and reinstall rbenv (including it's dependencies from ``` ~/.gem && ~/.rbenv ```) the try reinstalling this.  

also rbenv doesn't install bundler by default it is plausible that you are using system bundler to try and install nokogiri even though you are using rbenv's ruby.

 if you run ``` which bundler ``` and the return value is not rbenv something something try this: 

``` gem install bundler && rbenv rehash ```

**Comment by TimMoore** [comment]
I agree, this looks like filesystem permissions are set up incorrectly. Check permissions on your project directory and on `/Users/ratciti/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/nokogiri-1.6.6.2` (I believe it changes to the gem directory before building) to make sure you can write files as the user that you're running Bundler with.

**Comment by willbowling** [comment]
What resolved my issue was to install bundler without sudo. "gem install bundler".

**Comment by DemitryT** [comment]
:+1: for `gem install bundler`

**Comment by xicoofficial** [comment]
Installing bundle "gem install bundler" also solved my similar issue

**Comment by sergio-ocon** [comment]
I was using rbenv in Fedora and got the same message. I couldn't understand why.
At the end, it was a lack of underneath packages, it was solved installing them:
sudo dnf install ruby-devel cmake mariadb-devel rubygems-devel
bundler update
bundler install

**Comment by jclaret** [comment]
it works for fedora 22 too. 

thanks sergio-ocon!

**Comment by Viktorminator** [comment]
```
gem install bundler
```
solved the problem

**Comment by kmcmcurtis7** [comment]
Thanks for the info!  I ran into this and this resolved the problem quickly.

**Comment by Abdull** [comment]
`sudo apt-get install ruby2.2-dev`

**Comment by zoomwebstudio** [comment]
This solution works perfectly for me. thanks to @Abdull 
`sudo apt-get install ruby2.2-dev`
and after I ran 
`bundle install`

**Comment by shrinivas91** [comment]
I used sudo apt-get install ruby2.2-dev
But I'm getting the following error:
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package ruby2.2-dev
E: Couldn't find any package by regex 'ruby2.2-dev'

**Comment by dailsoncampos** [comment]
My problem is this "Make sure that `gem install debugger -v '1.6.8'` succeeds before bundling error". What can I do?

**Comment by willbowling** [comment]
Should probably run gem bundler install? If that is what you intend to use.
Ie "bundle exec" etc.

I also personally use rvm for my ruby versions. It seems more maintainable.

**Comment by dailsoncampos** [comment]
I run "bundle install".

➜  dailsoncampos git:(Production) ✗ bundle install                
Fetching gem metadata from https://rubygems.org/...........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 11.1.2
Using i18n 0.7.0
Using json 1.8.3
Using minitest 5.8.4
Using thread_safe 0.3.5
Using builder 3.2.2
Using erubis 2.7.0
Using mini_portile2 2.0.0
Using rack 1.6.4
Using mime-types 2.99
Using arel 6.0.3
Using debug_inspector 0.0.2
Using bundler 1.11.2
Using coffee-script-source 1.10.0
Using execjs 2.6.0
Using thor 0.19.1
Using columnize 0.9.0
Using concurrent-ruby 1.0.1
Using debugger-linecache 1.2.0
Using debugger-ruby_core_source 1.3.8
Using multi_json 1.11.2
Using sass 3.4.22
Using tilt 2.0.2
Using spring 1.6.4
Using sqlite3 1.3.11
Using rdoc 4.2.2
Using tzinfo 1.2.2
Using nokogiri 1.6.7.2
Using rack-test 0.6.3
Using mail 2.6.3
Using binding_of_caller 0.7.2
Using coffee-script 2.4.1
Using uglifier 2.7.2
Using sprockets 3.5.2
Installing debugger 1.6.8 with native extensions

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /home/campos/.rvm/rubies/ruby-2.1.8/bin/ruby -r ./siteconf20160423-10896-9be59d.rb extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/home/campos/.rvm/rubies/ruby-2.1.8/bin/ruby
/home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1401:in `initialize': No such file or directory @ rb_sysopen - ./218/ruby_debug.h (Errno::ENOENT)
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1401:in `open'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1401:in `copy_file'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:483:in `copy_file'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:400:in `block in cp'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1579:in `block in fu_each_src_dest'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1593:in `fu_each_src_dest0'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:1577:in `fu_each_src_dest'
	from /home/campos/.rvm/rubies/ruby-2.1.8/lib/ruby/2.1.0/fileutils.rb:399:in `cp'
	from extconf.rb:83:in `block in <main>'
	from extconf.rb:82:in `each'
	from extconf.rb:82:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /home/campos/.rvm/gems/ruby-2.1.8/gems/debugger-1.6.8 for inspection.
Results logged to /home/campos/.rvm/gems/ruby-2.1.8/extensions/x86_64-linux/2.1.0/debugger-1.6.8/gem_make.out
Using sdoc 0.4.1
Using activesupport 4.2.5.1
Using loofah 2.0.3
An error occurred while installing debugger (1.6.8), and Bundler cannot
continue.
Make sure that `gem install debugger -v '1.6.8'` succeeds before bundling.

**Comment by dailsoncampos** [comment]
I'm sorry. I'm beginner in ruby on rails...

**Comment by willbowling** [comment]
Try 'gem bundler install'. Even though when you run the command, it's
'bundle exec'. Bundler === bundle. Hope that helps.

**Comment by willbowling** [comment]
At work we use the combo "bundle exec && bundle compass watch" or "bundle
exec && bundle guard watch"

**Comment by dailsoncampos** [comment]
Man! Is there any problem if I comment the gem debugger? Cause I've commented and it's running...

**Comment by willbowling** [comment]
Can't help there man. Sounds like a gem I don't use. Good luck!

**Comment by dailsoncampos** [comment]
Thank you! I'm sorry my bad English :smile:

**Comment by ArionHardison** [comment]
```gem uninstall nokogiri
xcode-select --install
gem install nokogiri```

Worked for me.

**Comment by vikrantgithub** [comment]
[error.txt](https://github.com/bundler/bundler/files/500410/error.txt)
please help me

**Comment by supertopher** [comment]
@vikrantgithub this isn't really the place for this kind of question exactly.  given that this appears to be a heroku deployment you should ask that question in a heroku-related venue.  

Bundler is working as intended for your use case and this is a place for bundler questions.

that said however, heroku doesn't support `mysql`.  you will need to migrate to `postrges` or a database type that heroku does support. [This article may be helpful to you](https://devcenter.heroku.com/articles/getting-started-with-ruby#use-a-database) good luck

**Comment by shabbir-ahmed** [comment]
I run `bundle install`

Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing
your bundle as root will break this application for all non-root users on this
machine.
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 12.0.0
Using concurrent-ruby 1.0.4
Using i18n 0.7.0
Using minitest 5.10.1
Using thread_safe 0.3.5
Using builder 3.2.2
Using erubis 2.7.0
Using mini_portile2 2.1.0
Using rack 2.0.1
Installing nio4r 1.2.1 with native extensions
Using websocket-extensions 0.1.2
Using mime-types-data 3.2016.0521
Using arel 7.1.4
Installing debug_inspector 0.0.2 with native extensions
Using bundler 1.13.7
Installing byebug 9.0.6 with native extensions
Using coffee-script-source 1.12.2
Using execjs 2.7.0
Using method_source 0.8.2
Using thor 0.19.4
Using multi_json 1.12.1
Using json 1.8.3
Installing mysql2 0.4.5 with native extensions
Using rdoc 4.3.0
Using sass 3.4.23
Using tilt 2.0.5
Using turbolinks-source 5.0.0
Using tzinfo 1.2.2
Installing nokogiri 1.7.0.1 with native extensions
Using rack-test 0.6.3
Using sprockets 3.7.1
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /var/lib/gems/2.3.0/gems/nio4r-1.2.1/ext/nio4r
/usr/bin/ruby2.3 -r ./siteconf20170109-3150-tw7km1.rb extconf.rb
--with-cflags=-std=c99
mkmf.rb can't find header files for ruby at /usr/lib/ruby/include/ruby.h

extconf failed, exit code 1

Gem files will remain installed in /var/lib/gems/2.3.0/gems/nio4r-1.2.1 for
inspection.
Results logged to
/var/lib/gems/2.3.0/extensions/x86_64-linux/2.3.0/nio4r-1.2.1/gem_make.out

An error occurred while installing nio4r (1.2.1), and Bundler cannot
continue.
Make sure that `gem install nio4r -v '1.2.1'` succeeds before bundling.

I don't understanding what is it

**Comment by colby-swandale** [comment]
@shabbir-ahmed you're missing the required ruby headers to install the library. Look into installing `ruby-dev` or similar on your platform.

**Comment by ElijahLynn** [comment]
I had this issue by using a too new version of Ruby. In my case I was using 2.4.0 and when I used 2.2.0 it worked fine.

**Comment by Sihui** [comment]
My case was b/c I hadn't agreed to xcode license. 
Doing this made it work `sudo xcodebuild -license`

**Comment by vaibhavgeek** [comment]
`sudo apt-get install ruby-dev` 

solved the problem for me.

**Comment by anapitalai** [comment]
Perfect, a new version works.

**Comment by HuhMe9** [comment]
Grhdhfufufiefvrksu😋😄gfjrutufy🤗

بتاريخ ٠٦‏/٠٩‏/٢٠١٨ ١٢:٤٣، كتب "Alois Napitalai" <notifications@github.com>:

**Comment by odanicola** [comment]
@ArionHardison that worked for me, thanks man!

**Comment by Prashanth004** [comment]
@vaibhavgeek thanks it worked for me too
